systems/controllers/QPCommon.cpp

```diff
@@ -383,12 +427,98 @@ int setupAndSolveQP(NewQPControllerData *pdata, std::shared_ptr<drake::lcmt_qp_c
+
+    std::pair<Eigen::Vector3d, double> cop_and_normal_torque = pdata->r->resolveCenterOfPressure(force_torque_measurements, normal, point_on_contact_plane);
+    Vector2d zmp_from_force_sensors = cop_and_normal_torque.first.head<2>();
+
+    double dt;
+    if (pdata->state.t_prev == 0) {
+      pdata->state.center_of_mass_observer_state.topRows<2>() = xcom.topRows<2>();
+      pdata->state.center_of_mass_observer_state.bottomRows<2>() = xcomdot.topRows<2>();
+      dt = 0;
+    } else {
+      dt = robot_state.t - pdata->state.t_prev;
+    }
+    updateEstimatedCenterOfMassState(
+        zmp_from_force_sensors,
+        xcomdot.topRows<2>(),
+        xcom(2) - average_contact_point_height,
+        -pdata->r->a_grav(5),
+        dt,
+        pdata->state.last_com_ddot,
+        params->center_of_mass_observer_gain,
+        pdata->state.center_of_mass_observer_state);
+
+    if (DEBUG_ZMP_COM_OBSERVER) {
+    // publish zmp/com observer state
+//      std::cout << "dt: " << dt << std::endl;
+      std::unique_ptr<lcm::LCM> lcm_cpp(new lcm::LCM);
+     
+      drake::lcmt_zmp_com_observer_state zmp_com_observer_state_msg;
+      eigenVectorToCArray(pdata->state.center_of_mass_observer_state.head<2>(), zmp_com_observer_state_msg.com);
+      eigenVectorToCArray(pdata->state.center_of_mass_observer_state.tail<2>(), zmp_com_observer_state_msg.comd);
+      lcm_cpp->publish("ZMP_COM_OBSERVER_STATE", &zmp_com_observer_state_msg);
+
+      lcm_t * lcm;
+      lcm = lcm_create(NULL);
+      if (!lcm)
+        throw std::runtime_error("LCM not good");
+
+      double com_estimates_draw_height = average_contact_point_height + 0.01; // a little above the ground to make sure it doesn't get occluded by the ground plane.
+
+      bot_lcmgl_t* lcmgl = bot_lcmgl_init(lcm, "zmp com observer (zmp-based red, kinematic grey)");
+      bot_lcmgl_color3f(lcmgl, 1.0, 0.0, 0.0); // red
+      double observer_com_xyz[] = { zmp_com_observer_state_msg.com[0], zmp_com_observer_state_msg.com[1], com_estimates_draw_height };
+      bot_lcmgl_sphere(lcmgl, observer_com_xyz, .01, 36, 36);
+
+      bot_lcmgl_color3f(lcmgl, 0.65, 0.65, 0.65); // grey
+      double kinematics_com_xyz[] = {xcom[0], xcom[1], com_estimates_draw_height};
+      bot_lcmgl_sphere(lcmgl, kinematics_com_xyz, .01, 36, 36);
+
+      // zmp_from_force_sensors
+      bot_lcmgl_color3f(lcmgl, 0.0, 1.0, 0.0); // green
+      double zmp_from_force_sensors_xyz[] = {zmp_from_force_sensors[0], zmp_from_force_sensors[1], zmp_from_force_sensors[2]};
+      bot_lcmgl_sphere(lcmgl, zmp_from_force_sensors_xyz, .01, 36, 36);
+
+      bot_lcmgl_switch_buffer(lcmgl);
+      bot_lcmgl_destroy(lcmgl);
+      lcm_destroy(lcm);
+    }
+
+    // overwrite the com position and velocity with the observer's estimates:
+    xcom.topRows<2>() = pdata->state.center_of_mass_observer_state.topRows<2>();
+    // xcomdot.topRows<2>() = pdata->state.center_of_mass_observer_state.bottomRows<2>();
+  }
+
```

Comment:
can we refactor this out into a function? setupAndSolveQP is already ridiculously long